scripts/Modules/TTK_Lib.py

```diff
@@ -0,0 +1,355 @@
+    x8Mask = 0x0
+    if a and b and not prevMask in (0x0, 0x2, 0x3, 0x7):
+        x8Mask = 0x8
+    return int(a) + int(b) * 0x2 + int(l) * 0x4 + x8Mask
+    
+def encodeDirectionInput(x, y):
+    return (x << 4) + y
+    
+def encodeTrickInput(input):
+    return input * 0x10
+    
+# Reads binary data in-memory for the specified section
+def readRawRKGData(playerType: PlayerType, inputType: ControllerInputType) -> list:
+    retList = []
+    curAddr = 0x0
+    endAddr = 0x0
+    addresses = []
+    endAddrOffset = 0
+    
+    # Determine memory region to access
+    if (playerType == PlayerType.PLAYER):
+        addresses = classes.getInputStorageAddresses()
+        endAddrOffset = 0x276C
+    else:
+        addresses = classes.getGhostAddresses()
+        endAddrOffset = memory.read_u32(classes.getGhostAddressLengthPointer()[2])
+    
+    # Define the address range for the given inputType
+    curAddr = addresses[inputType.value]
+    if (inputType.value < 2):
+        endAddr = addresses[inputType.value + 1]
+    else:
+        endAddr = curAddr + endAddrOffset
+    
+    # Begin reading the data
+    dataTuple = memory.read_u16(curAddr)
+    while True:
+        retList.append(dataTuple >> 8)
+        retList.append(dataTuple % 0x100)
+        curAddr += 0x2
+        dataTuple = memory.read_u16(curAddr)
+        
+        if (dataTuple == 0x0000 or curAddr >= endAddr):
+            break
+    
+    return retList
+    
+# Expand raw rkg data into a list of frames
+def decodeRKGData(data: list, inputType: ControllerInputType) -> List[List[int]]:
+    retList = []
+    
+    if (inputType == ControllerInputType.TRICK):
+        trickInput = 0x0
+        x100Length = 0x0
+        
+        for i in range(0, len(data)):
+            dataByte = data[i]
+            
+            if (i %2) == 0:
+                trickInput = decodeTrickInput(dataByte)
+                x100Length = dataByte % 0x10
+            else:
+                dataLength = x100Length * 0x100 + dataByte
+                retList += [trickInput] * dataLength
+    else:
+        rawInput = 0x0
+        for i in range(0, len(data)):
+            dataByte = data[i]
+            
+            if (i %2) == 0:
+                rawInput = dataByte
+            else:
+                if (inputType == ControllerInputType.FACE):
+                    retList += [decodeFaceButton(rawInput)] * dataByte
+                else:
+                    retList += [list(map(lambda x: x-7, decodeDirectionInput(rawInput)))] * dataByte
+    return retList
+
+# Transform raw RKG data into a FrameSequence
+def readFullDecodedRKGData(playerType: PlayerType) -> Optional[FrameSequence]:
+    # First make sure we're actually in a race, otherwise we need to bail out
+    stage = classes.RaceInfo.stage()
+    if (stage == 0):
+        gui.add_osd_message("Not in race!")
+        return None
+    if (stage == 4):
+        gui.add_osd_message("Race is over!")
+        return None
```

Comment:
I'm of the opinion that we should only error out if the RKG data is unavailable, which is not the case here. This is especially applicable here as this is meant to be library code.

Reply:
After discussing with vabold, we found that the pointer chains are different during the intro camera and after finish. Further enhancement could be made to support these, but I currently think it's fine to require the user to be in-race in order to save race data to CSV/RKG. This PR is complicated enough anyway...